### PR TITLE
added support for Dict action space

### PR DESCRIPTION
**Pull request by JaCoderX**
Epsilon greedy policy error when generating new action from Dict Action space. #276

### CODE REVIEW

**Comment by JaCoderX** [comment]
@googlebot I signed it!

**Comment by tfboyd** [comment]
@oars I saw you did a +2 on the internal sync, so I added you as the reviewer here, you can assign someone else.

**Comment by ebrevdo** [comment]
We're trying to tighten up the contract that Agents present to the world, in particular we don't want to allow arbitrary nested structures "so long as its only got one entry", if the agent does not support more than one action in the action space.  Better if you just add a wrapper that converts the tensor to a dict.

If we start allowing this kind of soft contract, it makes it much harder for future developers to know which agents *really* support multiple action spaces; and makes it harder for us to add pytype to tf-agents.  Sorry; will have to reject this PR.

**Comment by ebrevdo** [comment]
(a good followup PR would be to disallow nested action_specs in this agent in the first place).

**Comment by JaCoderX** [comment]
> We're trying to tighten up the contract that Agents present to the world, in particular we don't want to allow arbitrary nested structures "so long as its only got one entry", if the agent does not support more than one action in the action space.

@ebrevdo, I agree with you that the code need to be simple and clean for the agents, especially if a certain agent implementation (this case DQN) is meant for just one dimensional action space.

but this PR contain 2 commits, one for the DQN and the second is for the epsilon greedy policy. 
As far as I understand, the epsilon greedy policy isn't bounded only for DQN and can be used by other agents that can have a structured action space? is that commit also too case specific?

personally I just wanted to test the tf-agent framework with an external and complex gym project and see they work well together. I use the DQN tutorial just for an easy entry to tf-agent framework.

anyway thanks for helping out :)

**Comment by ebrevdo** [comment]
Ah I didn't see the changes to epsilon_greedy.  That change is fine!  Can you revert the change to DQN agent?

File: tf_agents/agents/dqn/dqn_agent.py [hunk 1]
```diff
@@ -546,7 +547,8 @@ def _compute_next_q_values(self, next_time_steps):
     # action constraints are respected and helps centralize the greedy logic.
     greedy_actions = self._target_greedy_policy.action(
         next_time_steps, dummy_state).action
-
+    greedy_actions = tf.nest.flatten(greedy_actions)[0]
```
Comment (ebrevdo) [inline tf_agents/agents/dqn/dqn_agent.py]: revert this file please.

**Comment by JaCoderX** [comment]
DQN changes had been reverted

**Review by ebrevdo — APPROVED** [review]

**Comment by ebrevdo** [comment]
Is there a unit test?

**Comment by tfboyd** [comment]
@JacobHanouna  If there is a unit test, please reference it. If not can you add one and we can merge this can get it closed. Not to be a drag, but without a test the chances of it getting broken by some other change increases greatly.  Thank you for understanding.  We are **so close** on this and appreciate you powering through to the end.

**Comment by JaCoderX** [comment]
[Tutorial: Integration with TF-Agents RL Framework](https://github.com/Kismuz/btgym/issues/133)

**Comment by tfboyd** [comment]
So close to high fives all around.

**Comment by tfboyd** [comment]
MERGED

**Pull request merged**